controllers/machine_controller_test.go

```diff
@@ -32,11 +33,156 @@ import (
+	g.Expect(err).ShouldNot(HaveOccurred())
+	testCluster.Status.ControlPlaneInitialized = true
+	g.Expect(patchHelper.Patch(ctx, testCluster, patch.WithStatusObservedGeneration{})).ShouldNot(HaveOccurred())
+
+	// Patch infra machine ready
+	patchHelper, err = patch.NewHelper(infraMachine, testEnv)
+	g.Expect(err).ShouldNot(HaveOccurred())
+	g.Expect(unstructured.SetNestedField(infraMachine.Object, true, "status", "ready")).NotTo(HaveOccurred())
+	g.Expect(patchHelper.Patch(ctx, infraMachine, patch.WithStatusObservedGeneration{})).ShouldNot(HaveOccurred())
+
+	// Patch bootstrap ready
+	patchHelper, err = patch.NewHelper(defaultBootstrap, testEnv)
+	g.Expect(err).ShouldNot(HaveOccurred())
+	g.Expect(unstructured.SetNestedField(defaultBootstrap.Object, true, "status", "ready")).NotTo(HaveOccurred())
+	g.Expect(unstructured.SetNestedField(defaultBootstrap.Object, "secretData", "status", "dataSecretName")).To(Succeed())
+	g.Expect(patchHelper.Patch(ctx, defaultBootstrap, patch.WithStatusObservedGeneration{})).ShouldNot(HaveOccurred())
+
+	machine := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "machine-created-",
+			Namespace:    "default",
+			Finalizers:   []string{clusterv1.MachineFinalizer},
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName: testCluster.Name,
+			InfrastructureRef: corev1.ObjectReference{
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+				Kind:       "InfrastructureMachine",
+				Name:       "infra-config1",
+			},
+			Bootstrap: clusterv1.Bootstrap{
+				ConfigRef: &corev1.ObjectReference{
+					APIVersion: "bootstrap.cluster.x-k8s.io/v1alpha3",
+					Kind:       "BootstrapMachine",
+					Name:       "bootstrap-config-machinereconcile",
+				},
+			}},
+	}
+
+	g.Expect(testEnv.Create(ctx, machine)).To(BeNil())
+
+	// Wait for reconciliation to happen.
+	// Since infra and bootstrap objects are ready, a nodeRef will be assigned during nodereconcilation.
+	key := client.ObjectKey{Name: machine.Name, Namespace: machine.Namespace}
+	g.Eventually(func() bool {
+		if err := testEnv.Get(ctx, key, machine); err != nil {
+			return false
+		}
+		return machine.Status.NodeRef != nil
+	}, timeout).Should(BeTrue())
+
+	g.Consistently(func() bool { return machine.Status.NodeRef != nil }, 2*time.Second, 100*time.Millisecond).Should(BeTrue())
+
+	// Node deletion will trigger node watchers and a request will be added to the queue.
+	//
+	g.Expect(testEnv.Delete(ctx, node)).NotTo(HaveOccurred())
+	// TODO: Once conditions are in place, check if node deletion triggered a reconcile.
+
+	// Delete infra machine, external tracker will trigger reconcile
+	// and machine Status.FailureReason should be non-nil after reconcileInfrastructure
+	g.Expect(testEnv.Delete(ctx, infraMachine)).NotTo(HaveOccurred())
+	g.Eventually(func() bool {
+		if err := testEnv.Get(ctx, key, machine); err != nil {
+			return false
+		}
+		return machine.Status.FailureMessage != nil
+	}, timeout).Should(BeTrue())
```

Comment:
Why would we need to delete the infrastructure machine? Is this for cleanup purposes?

Reply:
This test is to check if external watches and node watches correctly trigger reconciliation on change or not. Renamed to TestWatches()